File: outputDir/vscripts/features/NullSafetyDemo.lua
-- package: com.example.features

require("ktox-lib")
ktox_sourcemap_traceback(debug and debug.getinfo and (debug.getinfo(1) or {}).short_src or "", "features/NullSafetyDemo.kt", {["1-16"]=1,["17"]=20,["18-20"]=21,["21-31"]=23,["32"]=24,["33"]=25,["34"]=26,["35-38"]=27,["39"]=33,["40-47"]=34,["48-55"]=37,["56-63"]=40,["64-71"]=43,["72-79"]=44,["80"]=47,["81-88"]=48,["89-91"]=50})
ktox_require("math/Calculator")

NullSafetyDemo = {}
NullSafetyDemo.__index = NullSafetyDemo

function NullSafetyDemo:new()
    local self = setmetatable({}, NullSafetyDemo)
    return self
end


function demoNullSafety()
    local calc = nil
    if 1 > 0.0 then
        calc = Calculator:new()
    end
    local testMe = function()
        do
            local __t0 = calc
            local __t1 = (__t0 ~= nil) and __t0:add(3, 7) or nil
            if __t1 ~= nil then
                return __t1
            else
                return -1
            end
        end
    end
    local resultA = testMe()
    calc = nil
    local resultB = testMe()
    return resultA * resultB
end

function demoScopeFunctions()
    local base = 5
    local logged
    do
        local __t0 = base
        local __t1 = (__t0 ~= nil) and ktox_also(__t0, function(it)
            return println(it)
        end) or nil
        logged = __t1
    end
    local doubled
    do
        local __t0 = logged
        local __t1 = (__t0 ~= nil) and ktox_let(__t0, function(it)
            return it * 2
        end) or nil
        doubled = __t1
    end
    local tripled
    do
        local __t0 = doubled
        local __t1 = (__t0 ~= nil) and ktox_let(__t0, function(n)
            return n + doubled
        end) or nil
        tripled = __t1
    end
    local positive
    do
        local __t0 = tripled
        local __t1 = (__t0 ~= nil) and ktox_takeIf(__t0, function(it)
            return it > 0
        end) or nil
        positive = __t1
    end
    local negative
    do
        local __t0 = tripled
        local __t1 = (__t0 ~= nil) and ktox_takeIf(__t0, function(it)
            return it < 0
        end) or nil
        negative = __t1
    end
    local maybeNull = nil
    local safe
    do
        local __t0 = maybeNull
        local __t1 = (__t0 ~= nil) and ktox_also(__t0, function(it)
            return println(it)
        end) or nil
        safe = __t1
    end
    return (ktox_elvis(positive, 0)) + (ktox_elvis(negative, -1)) + ((safe == nil and 0 or 1))
end

File: src/main/kotlin/com/example/features/NullSafetyDemo.kt
package com.example.features

import com.example.math.Calculator

/**
 * Demonstrates null-safe chain desugaring (`?.` and `?:`) and Kotlin scope
 * functions (`also`, `let`, `takeIf`).
 *
 * Each `?.` step becomes a `do...end` block with `local __tN` temporaries:
 *   - receiver evaluated once into `__t0`
 *   - each step: `local __tN = (__t{N-1} ~= nil) and <access> or nil`
 *   - Elvis (`?:`) adds an `if __tFinal ~= nil then … else … end`
 *
 * `demoNullSafety()` and `demoScopeFunctions()` are global functions accessible
 * from Main.lua after `require("features/NullSafetyDemo")`.
 */
class NullSafetyDemo

fun demoNullSafety(): Int {
    var calc: Calculator? = null
    if (1 > 0.0) calc = Calculator()
    // calc is not nil → __t0:add(3, 7) == 10, Elvis default never taken
    val testMe = { calc?.add(3, 7) ?: -1 }
    val resultA = testMe()
    calc = null
    val resultB = testMe()
    return resultA * resultB
}

/** Demonstrates `also`, `let`, and `takeIf` scope functions. */
fun demoScopeFunctions(): Int {
    // also: executes block with the receiver, returns the receiver
    val base: Int? = 5
    val logged = base?.also { println(it) }  // logged == 5

    // let: transforms the value
    val doubled = logged?.let { it * 2 }     // doubled == 10

    // let with explicit parameter name
    val tripled = doubled?.let { n -> n + doubled }  // tripled == 20

    // takeIf: returns receiver when predicate is true, else null
    val positive = tripled?.takeIf { it > 0 }  // positive == 20
    val negative = tripled?.takeIf { it < 0 }  // negative == null

    // null-safe chain with scope function
    val maybeNull: Int? = null
    val safe = maybeNull?.also { println(it) }  // safe == null

    return (positive ?: 0) + (negative ?: -1) + (if (safe == null) 0 else 1)
}
